### PR TITLE
Add validation to blocklists, test that adding doesn't break stuff

### PR DESCRIPTION
**Pull request by RangerMauve**
Related to #37 where there's an error being swallowed and leading to unexpected behavior.

### CODE REVIEW

File: src/server/api/blockallowlist.ts [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
 import ActivityPubSystem from '../apsystem.js'
 
+export function validateBlocklistFormat (mention: string): void {
+  const sections = mention.split('@')
+  const correctSections = sections.length === 3
+  const startsEmpty = sections[0] === ''
+  const hasUserAndDomain = (sections[1] !== '') && (sections[2] !== '')
+  const isValid = correctSections && hasUserAndDomain && startsEmpty
+  if (!isValid) {
+    throw createError(400, `Invalid account format: ${mention}. Please use the syntax @usernam@domain`)
```
Comment (catdevnull) [inline src/server/api/blockallowlist.ts:16]:
```suggestion
    throw createError(400, `Invalid account format: ${mention}. Please use the syntax @username@domain`)
```

**Comment by catdevnull** [comment]
where are blocklists from? is it possible they may include `# comments` or empty lines or spaces before and after a username?

**Review by catdevnull — APPROVED** [review]

File: src/server/api/blockallowlist.ts [hunk 2]
```diff
@@ -45,7 +58,8 @@ export const blockAllowListRoutes = (cfg: APIConfig, store: Store, apsystem: Act
       return await reply.code(403).send('Not Allowed')
     }
 
-    const accounts = request.body.split('\n')
+    const accounts = request.body.trim().split('\n')
```
Comment (fauno) [inline src/server/api/blockallowlist.ts:61]: maybe `.map(x => x.trim())` after the split? also remove empty lines

File: src/server/api/blockallowlist.ts [hunk 3]
```diff
@@ -1,9 +1,22 @@
...
+import createError from 'http-errors'
+
 import Store from '../store/index.js'
 import ActivityPubSystem from '../apsystem.js'
 
+export function validateBlocklistFormat (mention: string): void {
+  const sections = mention.split('@')
+  const correctSections = sections.length === 3
+  const startsEmpty = sections[0] === ''
+  const hasUserAndDomain = (sections[1] !== '') && (sections[2] !== '')
```
Comment (fauno) [inline src/server/api/blockallowlist.ts:13]: blocking a whole instance is `@*@doma.in` no?

**Comment by fauno** [comment]
> where are blocklists from? is it possible they may include `# comments` or empty lines or spaces before and after a username?

the api client sends accounts separated by newlines to the endpoint. on the ruby client the method takes an array and concatenates items with a newline

File: src/server/api/blockallowlist.ts (hunk 3 above)
Comment (RangerMauve) [reply src/server/api/blockallowlist.ts:13]: Yup! In that cause username is set to `*`. `@@domain.com` would be invalid for example.

File: src/server/api/blockallowlist.ts [hunk 4]
```diff
@@ -1,9 +1,22 @@
 import { APIConfig, FastifyTypebox } from '.'
 import { Type } from '@sinclair/typebox'
 
+import createError from 'http-errors'
+
 import Store from '../store/index.js'
 import ActivityPubSystem from '../apsystem.js'
 
+export function validateBlocklistFormat (mention: string): void {
+  const sections = mention.split('@')
```
Comment (akhileshthite) [inline src/server/api/blockallowlist.ts:10]:
```javascript
const isInstanceBlock = sections[1] === '*' && sections[0] === '' && sections[2] !== '';
```

**Comment by RangerMauve** [comment]
Yeah I'd like to avoid adding any fancy syntax like comments or empty lines and stick to valid user input. I added trim to account for newlines at the end messing stuff up since it'd be a likely error when importing a file.

File: src/server/api/blockallowlist.ts (hunk 2 above)
Comment (RangerMauve) [reply src/server/api/blockallowlist.ts:61]: I think empty lines should be an error.

**Review by akhileshthite — APPROVED** [review]
closes #37

**Review by fauno — APPROVED** [review]
I'm ok with this even if each item isn't trimmed

File: src/server/api/blockallowlist.ts (hunk 4 above)
Comment (RangerMauve) [reply src/server/api/blockallowlist.ts:10]: I don't think we need to explicitly handle this case since it's already handled by the other checks.

**Pull request merged**